File: brainfuck+swift/Sources/Brainfuck/Brainfuck.swift
enum Brainfuck {
    static func run(_ source: String, input: String = "") -> String? {
        guard let instructions = parse(source) else {
            return nil
        }
        let memory = Memory()
        var output = String.init()
        var inputIdx = input.startIndex
        
        var pc = 0
        loop: while (pc < instructions.count) {
            switch instructions[pc] {
            case .moveRight:
                memory.moveRight()
            case .moveLeft:
                memory.moveLeft()
            case .increment:
                memory.increment()
            case .decrement:
                memory.decrement()
            case .output:
                output.append(Character(UnicodeScalar(memory.get())))
            case .input:
                if inputIdx == input.endIndex {
                    memory.put(0)
                } else {
                    memory.put(input[inputIdx].asciiValue!)
                    inputIdx = input.index(after: inputIdx)
                }
            case .open(let j):
                if memory.get() == 0 {
                    pc += j
                    continue loop
                }
            case .close(let j):
                if memory.get() != 0 {
                    pc -= j
                    continue loop
                }
            }
            pc += 1
        }
    return output
    }
}

File: brainfuck+elixir/lib/brainfuck.ex
defmodule Brainfuck do
  alias Brainfuck.{Instruction, Memory}

  def run(source, input \\ "") do
    instructions =
      source
      |> Instruction.parse()
      |> List.to_tuple()

    Stream.resource(
      fn -> {0, Memory.new(), String.to_charlist(input)} end,
      fn {pc, memory, chars} ->
        if pc == tuple_size(instructions) do
          {:halt, :done}
        else
          case elem(instructions, pc) do
            :move_right ->
              {[], {pc + 1, Memory.move_right(memory), chars}}

            :move_left ->
              {[], {pc + 1, Memory.move_left(memory), chars}}

            :increment ->
              {[], {pc + 1, Memory.increment(memory), chars}}

            :decrement ->
              {[], {pc + 1, Memory.decrement(memory), chars}}

            :output ->
              {[Memory.get(memory)], {pc + 1, memory, chars}}

            :input ->
              {c, rest} =
                case chars do
                  [] -> {0, []}
                  [c] -> {c, []}
                  [c | rest] -> {c, rest}
                end

              {[], {pc + 1, Memory.put(memory, c), rest}}

            {:open, jump} ->
              if Memory.get(memory) == 0 do
                {[], {pc + jump, memory, chars}}
              else
                {[], {pc + 1, memory, chars}}
              end

            {:close, jump} ->
              if Memory.get(memory) != 0 do
                {[], {pc - jump, memory, chars}}
              else
                {[], {pc + 1, memory, chars}}
              end
          end
        end
      end,
      fn _ -> nil end
    )
    |> Enum.to_list()
    |> to_string()
  end
end
